tests/dualtor/test_tunnel_memory_leak.py

```diff
@@ -0,0 +1,183 @@
+
+def check_memory_leak(duthost):
+    """Check if it has memory leak on duthost
+
+    Args:
+        duthost (AnsibleHost): Device Under Test (DUT)
+
+    Returns:
+        bool: True if there is memory leak. Otherwise, return False.
+    """
+    global swss_mem_percent
+    stdout_lines = duthost.command("docker stats swss --no-stream")["stdout_lines"]
+    header = stdout_lines[0]
+    # Find the position of category "MEM USAGE", "MEM %" and "NET I/O"
+    pos1 = header.index("MEM USAGE")
+    pos2 = header.index("MEM %")
+    pos3 = header.index("NET I/O")
+    if len(stdout_lines) < 2:
+        pytest.fail("Collect swss stat failed, swss container may die.")
+    line = stdout_lines[-1]
+    # Get the value of "MEM USAGE", "LIMIT" and "MEM %"
+    mem_info = line[pos1:pos2].strip().split("/")
+    mem_usage = mem_info[0].strip()
+    mem_limit = mem_info[1].strip()
+    mem_percent = line[pos2:pos3].strip()
+
+    logging.info("SWSS MEM USAGE:{} LIMIT:{} PERCENT:{}".format(mem_usage, mem_limit, mem_percent))
+
+    mem_percent = float(mem_percent[:-1])
+    if not swss_mem_percent:
+        # Save swss mem usage at the first time.
+        swss_mem_percent = mem_percent
+        logging.info("SWSS container original MEM USAGE:{} original percent: {}%"
+                    .format(mem_usage, swss_mem_percent))
+        return False
+    elif mem_percent > swss_mem_percent:
+        logging.error("SWSS container MEM percent is increased. current percent:{}%, original percent: {}%"
+                    .format(mem_percent, swss_mem_percent))
+        return True
+    return False
+
+def test_tunnel_memory_leak(
+    toggle_all_simulator_ports_to_upper_tor,
+    upper_tor_host, lower_tor_host, ptfhost, 
+    ptfadapter, conn_graph_facts, tbinfo, vmhost
+):
+    """
+    Test if there is memory leak for service tunnel_packet_handler.
+    Send ip packets from standby TOR T1 to Server, standby TOR will
+    forward the packets to active TOR with tunnel, active TOR will
+    decapsulate the IPinIP packets, but there is no neighbor for destination
+    as we remove neighbor before test, tunnel_packet_handler will be
+    triggered and neighbor will be added. Server will receive the packets.
+    Check if memory usage is increased after tunnel_packet_handler's
+    operation. Since tunnel_packet_handler is only triggered by the
+    first packet, loop the process for all severs to trigger it as much
+    as possible.
+    """
+    @contextlib.contextmanager
+    def prepare_services(ptfhost):
+        """
+        Temporarily stop garp service, start arp and icmp service.
+        It has to stop garp_service for triggering tunnel_packet_handler.
+        It has to start arp and icmp service for receiving packets at server side. 
+        """
+        ptfhost.shell("supervisorctl stop garp_service")
+        ptfhost.shell("supervisorctl start arp_responder")
+        ptfhost.shell("supervisorctl start icmp_responder")
```

Comment:
These services should be stopped, since the tunnel packet handler operates in the scenario where no ARP information is learned for a given neighbor. If they are started, `validate_neighbor_entry_exists` should always return true since `arp_responder` will allow the neighbor to be learned normally.

Reply:
Before test, I start `arp_responder `and `icmp_responder`, the reason is I want to make sure server can receive the packets.
Before sending packet, I delete neighbor entry for the server ip, this step is to make sure tunnel_packet_handler will be triggered, and I have verified it on testbed, `arp_responder `will not add neighbor entry back after we delete it, `garp_responder `will, but we stop it.
I checked the log, tunnel_packet_handler is triggered for every server in the test process.

Reply:
Looks like I misunderstood the test case, sorry about that. This looks good.